// @sap/bas-sdk/dist/test/apis/ai.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const sinon_1 = __importDefault(require("sinon"));
const proxyquire_1 = __importDefault(require("proxyquire"));
describe("sendAiRequest", () => {
    let sendAiRequest;
    let verifyDefaultLandscapeStub;
    let sendRequestStub;
    let isAppStudioStub;
    let getTenantMetadataStub;
    beforeEach(() => {
        verifyDefaultLandscapeStub = sinon_1.default.stub();
        sendRequestStub = sinon_1.default.stub();
        getTenantMetadataStub = sinon_1.default.stub();
        isAppStudioStub = sinon_1.default.stub();
        const { sendAiRequest: aiRequestFunc } = (0, proxyquire_1.default)("../../src/apis/ai", {
            "../utils/devspace-utils": {
                sendRequest: sendRequestStub,
                verifyDefaultLandscape: verifyDefaultLandscapeStub,
            },
            "./get-devspace": { getTenantMetadata: getTenantMetadataStub },
            "./is-app-studio": { isAppStudio: isAppStudioStub },
        });
        sendAiRequest = aiRequestFunc;
    });
    afterEach(() => {
        sinon_1.default.restore();
    });
    const tenantMetadata = {
        host: "test-tenant-url",
        wsPlan: "test-ws-plan",
        buildCodePlan: "test-build-code-plan",
    };
    it("not BAS, should successfully send an ai request", async () => {
        // Setup
        const mockLandscape = "test-landscape";
        const mockJwt = "test-jwt";
        const mockRequest = {
            path: "/test-path",
            method: "GET",
            config: {
                data: { test: "data" },
                headers: { "Custom-Header": "Value" },
                custom: { test: "custom" },
            },
        };
        const mockResponse = { success: true };
        verifyDefaultLandscapeStub.resolves({
            jwt: mockJwt,
            landscape: mockLandscape,
        });
        getTenantMetadataStub.resolves(tenantMetadata);
        sendRequestStub.resolves(mockResponse);
        isAppStudioStub.returns(false);
        // Execute
        const result = await sendAiRequest(mockRequest);
        // Assert
        (0, chai_1.expect)(result).to.deep.equal(mockResponse);
        (0, chai_1.expect)(sendRequestStub.calledOnce).to.be.true;
        (0, chai_1.expect)(sendRequestStub.firstCall.args[0]).to.deep.include({
            jwt: mockJwt,
            config: Object.assign(Object.assign({}, mockRequest.config), { headers: Object.assign(Object.assign({}, mockRequest.config.headers), { "AI-Resource-Group": "default" }) }),
            method: mockRequest.method,
            url: {
                domain: `https://${tenantMetadata.host}`,
                path: `secure-outbound-connectivity/llm${mockRequest.path}`,
            },
        });
    });
    it("not BAS, should successfully send an ai request with custom resource group", async () => {
        // Setup
        const mockLandscape = "test-landscape";
        const mockJwt = "test-jwt";
        const mockRequest = {
            path: "/test-path",
            method: "GET",
            config: {
                data: { test: "data" },
                headers: { "Custom-Header": "Value", "AI-Resource-Group": "custom-rg" },
                custom: { test: "custom" },
            },
        };
        const mockResponse = { success: true };
        verifyDefaultLandscapeStub.resolves({
            jwt: mockJwt,
            landscape: mockLandscape,
        });
        getTenantMetadataStub.resolves(tenantMetadata);
        sendRequestStub.resolves(mockResponse);
        isAppStudioStub.returns(false);
        // Execute
        const result = await sendAiRequest(mockRequest);
        // Assert
        (0, chai_1.expect)(result).to.deep.equal(mockResponse);
        (0, chai_1.expect)(sendRequestStub.calledOnce).to.be.true;
        (0, chai_1.expect)(sendRequestStub.firstCall.args[0]).to.deep.include({
            jwt: mockJwt,
            config: Object.assign(Object.assign({}, mockRequest.config), { headers: Object.assign(Object.assign({}, mockRequest.config.headers), { "AI-Resource-Group": "custom-rg" }) }),
            method: mockRequest.method,
            url: {
                domain: `https://${tenantMetadata.host}`,
                path: `secure-outbound-connectivity/llm${mockRequest.path}`,
            },
        });
    });
    it("not BAS, should throw an error if no default landscape is set", async () => {
        isAppStudioStub.returns(false);
        verifyDefaultLandscapeStub.rejects(new Error("There is no default landscape set"));
        await (0, chai_1.expect)(sendAiRequest({ path: "/test", method: "GET" })).to.be.rejectedWith("There is no default landscape set");
    });
    it("is BAS, should successfully send an ai request to H2O domain", async () => {
        // Setup
        const mockRequest = {
            path: "/test-path",
            method: "GET",
        };
        const mockResponse = { success: true };
        sendRequestStub.resolves(mockResponse);
        // Stubbing process.env
        const h2oValue = "https://h2o.test.com";
        sinon_1.default.stub(process, "env").value({ H2O_URL: h2oValue });
        isAppStudioStub.returns(true);
        // Execute
        const result = await sendAiRequest(mockRequest);
        // Assert
        (0, chai_1.expect)(result).to.deep.equal(mockResponse);
        (0, chai_1.expect)(sendRequestStub.calledOnce).to.be.true;
        (0, chai_1.expect)(sendRequestStub.firstCall.args[0]).to.deep.include({
            config: {
                headers: {
                    "AI-Resource-Group": "default",
                },
            },
            method: mockRequest.method,
            url: {
                domain: h2oValue,
                path: `llm/test-path`,
            },
        });
    });
    it("is BAS, should successfully send an ai request to H2O domain, when the var is not populated", async () => {
        // Setup
        const mockRequest = {
            path: "/test-path",
            method: "GET",
            config: {
                data: { test: "data" },
                headers: { "Custom-Header": "Value" },
                responseType: "stream",
            },
        };
        const mockResponse = { success: true };
        sendRequestStub.resolves(mockResponse);
        // Stubbing process.env
        sinon_1.default.stub(process, "env").value({});
        isAppStudioStub.returns(true);
        // Execute
        const result = await sendAiRequest(mockRequest);
        // Assert
        (0, chai_1.expect)(result).to.deep.equal(mockResponse);
        (0, chai_1.expect)(sendRequestStub.calledOnce).to.be.true;
        (0, chai_1.expect)(sendRequestStub.firstCall.args[0]).to.deep.include({
            config: Object.assign(Object.assign({}, mockRequest.config), { headers: Object.assign(Object.assign({}, mockRequest.config.headers), { "AI-Resource-Group": "default" }) }),
            method: mockRequest.method,
            url: {
                domain: "",
                path: `llm/test-path`,
            },
        });
    });
});
