// @sap-ux/mockserver-config-writer/dist/i18n.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initI18n = initI18n;
exports.t = t;
const i18next_1 = __importDefault(require("i18next"));
const mockserver_config_writer_i18n_json_1 = __importDefault(require("./translations/mockserver-config-writer.i18n.json"));
const NS = 'mockserver-config-writer';
let i18nInstance;
/**
 * Initialize i18next with the translations for this module.
 */
async function initI18n() {
    i18nInstance = i18next_1.default.createInstance({
        resources: {
            en: {
                [NS]: mockserver_config_writer_i18n_json_1.default
            }
        },
        lng: 'en',
        fallbackLng: 'en',
        defaultNS: NS,
        ns: [NS]
    });
    await i18nInstance.init();
}
/**
 * Helper function facading the call to i18next.
 *
 * @param key i18n key
 * @param options additional options
 * @returns {string} localized string stored for the given key
 */
function t(key, options) {
    return i18nInstance.t(key, options);
}
initI18n().catch(() => {
    // Ignore any errors since the write will still work
});
