// @sap/bas-sdk/src/apis/create-destination.ts
import urljoin from "url-join";
import axios, { AxiosRequestConfig } from "axios";
import { getEnvValue } from "../utils/core-utils";
import { addJwtHeader } from "./headers-helper";
import {
  DestinationInfo,
  destinationInfoToAnswerDestination,
  ENV_H2O_URL,
} from "../utils/destinations";

/**
 * add new destination
 */
export async function createDestination(
  destinaionInfo: DestinationInfo,
): Promise<void> {
  const flatDestination = destinationInfoToAnswerDestination(destinaionInfo);
  const reqURL = urljoin(getEnvValue(ENV_H2O_URL), "/api/createDestination");
  try {
    /* Axios exposes the version in the user-agent header, creating a security issue. 
    To address this, we override the header for enhanced security.*/
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "User-Agent": "Bas",
    };

    const reqConfig: AxiosRequestConfig = {
      method: "post",
      url: reqURL,
      headers: addJwtHeader(headers),
      data: flatDestination,
    };
    await axios.request(reqConfig);
  } catch (err) {
    throw new Error(
      `Couldn't create the destination from the (${reqURL}) URL because of the following error: ${err.message}. Error code=${err.response.status}`,
    );
  }
}
