// @sap-ux/ui5-config/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fioriToolsProxy = exports.YAMLError = exports.yamlErrorCode = exports.replaceEnvVariables = exports.getTypesPackage = exports.getTypesVersion = exports.getEsmTypesVersion = exports.mergeObjects = exports.UI5Config = void 0;
var ui5config_1 = require("./ui5config");
Object.defineProperty(exports, "UI5Config", { enumerable: true, get: function () { return ui5config_1.UI5Config; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "mergeObjects", { enumerable: true, get: function () { return utils_1.mergeObjects; } });
Object.defineProperty(exports, "getEsmTypesVersion", { enumerable: true, get: function () { return utils_1.getEsmTypesVersion; } });
Object.defineProperty(exports, "getTypesVersion", { enumerable: true, get: function () { return utils_1.getTypesVersion; } });
Object.defineProperty(exports, "getTypesPackage", { enumerable: true, get: function () { return utils_1.getTypesPackage; } });
Object.defineProperty(exports, "replaceEnvVariables", { enumerable: true, get: function () { return utils_1.replaceEnvVariables; } });
var yaml_1 = require("@sap-ux/yaml");
Object.defineProperty(exports, "yamlErrorCode", { enumerable: true, get: function () { return yaml_1.errorCode; } });
Object.defineProperty(exports, "YAMLError", { enumerable: true, get: function () { return yaml_1.YAMLError; } });
var constants_1 = require("./constants");
Object.defineProperty(exports, "fioriToolsProxy", { enumerable: true, get: function () { return constants_1.fioriToolsProxy; } });
