// @sap-ux/preview-middleware/dist/client/thirdparty/@sap-ux-private/control-property-editor-common.js
sap.ui.define((function () { 'use strict';

	var commonjsGlobal = typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};

	var dist = {};

	var api = {};

	(function (exports) {
		Object.defineProperty(exports, "__esModule", { value: true });
		exports.requestControlContextMenu = exports.showInfoCenterMessage = exports.externalFileChange = exports.setApplicationRequiresReload = exports.executeContextMenuAction = exports.executeQuickAction = exports.updateQuickAction = exports.quickActionListChanged = exports.save = exports.redo = exports.undo = exports.appLoaded = exports.setSaveEnablement = exports.setUndoRedoEnablement = exports.applicationModeChanged = exports.setAppMode = exports.storageFileChanged = exports.reloadApplication = exports.showMessage = exports.changeStackModified = exports.propertyChangeFailed = exports.propertyChanged = exports.changeProperty = exports.outlineChanged = exports.deletePropertyChanges = exports.addExtensionPoint = exports.selectControl = exports.controlSelected = exports.iconsLoaded = exports.EXTERNAL_ACTION_PREFIX = exports.REJECTED_SUFFIX = exports.FULFILLED_SUFFIX = exports.PENDING_SUFFIX = exports.MessageBarType = exports.NESTED_QUICK_ACTION_KIND = exports.SIMPLE_QUICK_ACTION_KIND = exports.CONTROL_CHANGE_KIND = exports.GENERIC_CHANGE_KIND = exports.UNKNOWN_CHANGE_KIND = exports.SAVED_CHANGE_TYPE = exports.PENDING_CHANGE_TYPE = exports.PropertyType = exports.SCENARIO = exports.CHECKBOX_EDITOR_TYPE = exports.DROPDOWN_EDITOR_TYPE = exports.INPUT_EDITOR_TYPE = exports.STRING_VALUE_TYPE = exports.FLOAT_VALUE_TYPE = exports.INTEGER_VALUE_TYPE = exports.BOOLEAN_VALUE_TYPE = void 0;
		exports.createAsyncActionFactory = createAsyncActionFactory;
		exports.BOOLEAN_VALUE_TYPE = 'boolean';
		exports.INTEGER_VALUE_TYPE = 'integer';
		exports.FLOAT_VALUE_TYPE = 'float';
		exports.STRING_VALUE_TYPE = 'string';
		exports.INPUT_EDITOR_TYPE = 'input';
		exports.DROPDOWN_EDITOR_TYPE = 'dropdown';
		exports.CHECKBOX_EDITOR_TYPE = 'checkbox';
		exports.SCENARIO = {
		    AppVariant: 'APP_VARIANT',
		    VersionedAppVariant: 'VERSIONED_APP_VARIANT',
		    AdaptationProject: 'ADAPTATION_PROJECT',
		    FioriElementsFromScratch: 'FE_FROM_SCRATCH',
		    UiAdaptation: 'UI_ADAPTATION'
		};
		var PropertyType;
		(function (PropertyType) {
		    PropertyType["Configuration"] = "configuration";
		    PropertyType["ControlProperty"] = "controlProperty";
		})(PropertyType || (exports.PropertyType = PropertyType = {}));
		exports.PENDING_CHANGE_TYPE = 'pending';
		exports.SAVED_CHANGE_TYPE = 'saved';
		exports.UNKNOWN_CHANGE_KIND = 'unknown';
		exports.GENERIC_CHANGE_KIND = 'generic';
		exports.CONTROL_CHANGE_KIND = 'control';
		exports.SIMPLE_QUICK_ACTION_KIND = 'simple';
		exports.NESTED_QUICK_ACTION_KIND = 'nested';
		var MessageBarType;
		(function (MessageBarType) {
		    /** Info styled MessageBar */
		    MessageBarType[MessageBarType["info"] = 0] = "info";
		    /** Error styled MessageBar */
		    MessageBarType[MessageBarType["error"] = 1] = "error";
		    /** Warning styled MessageBar */
		    MessageBarType[MessageBarType["warning"] = 5] = "warning";
		})(MessageBarType || (exports.MessageBarType = MessageBarType = {}));
		/**
		 * Create matcher.
		 *
		 * @param type action type
		 * @returns (value: { type: unknown } | undefined) => value is Y
		 */
		function createMatcher(type) {
		    return function match(value) {
		        return value?.type === type;
		    };
		}
		/**
		 * Create action factory.
		 *
		 * @param prefix to determine ext action
		 * @returns Function
		 */
		function createActionFactory(prefix) {
		    return function createAction(name) {
		        const actionType = [prefix, name].join(' ');
		        /**
		         *
		         * @param payload action payload
		         * @returns PayloadAction<typeof actionType, T>
		         */
		        function action(payload) {
		            return {
		                type: actionType,
		                payload
		            };
		        }
		        action.type = actionType;
		        action.match = createMatcher(actionType);
		        return action;
		    };
		}
		exports.PENDING_SUFFIX = '<pending>';
		exports.FULFILLED_SUFFIX = '<fulfilled>';
		exports.REJECTED_SUFFIX = '<rejected>';
		/**
		 * Factory for creating  request response actions.
		 *
		 * @param prefix action prefix
		 * @returns Function
		 */
		function createAsyncActionFactory(prefix) {
		    return function createAction(name) {
		        const pendingType = [prefix, name, exports.PENDING_SUFFIX].join(' ');
		        /**
		         * Pending action.
		         *
		         * @param payload action payload
		         * @returns PayloadAction<typeof pendingType, T>
		         */
		        function pending(payload) {
		            return {
		                type: pendingType,
		                payload
		            };
		        }
		        pending.type = pendingType;
		        pending.match = createMatcher(pendingType);
		        const fulfilledType = [prefix, name, exports.FULFILLED_SUFFIX].join(' ');
		        /**
		         * Fulfill action.
		         *
		         * @param payload action payload
		         * @returns PayloadAction<typeof fulfilledType, F>
		         */
		        function fulfilled(payload) {
		            return {
		                type: fulfilledType,
		                payload
		            };
		        }
		        fulfilled.type = fulfilledType;
		        fulfilled.match = createMatcher(fulfilledType);
		        const rejectedType = [prefix, name, exports.REJECTED_SUFFIX].join(' ');
		        /**
		         * Reject action.
		         *
		         * @param message error message
		         * @param payload R
		         * @returns ErrorAction<typeof rejectedType, F>
		         */
		        function rejected(message, payload) {
		            return {
		                type: rejectedType,
		                payload,
		                error: {
		                    message
		                },
		                showMessage: true
		            };
		        }
		        rejected.type = rejectedType;
		        rejected.match = createMatcher(rejectedType);
		        return {
		            pending,
		            fulfilled,
		            rejected
		        };
		    };
		}
		exports.EXTERNAL_ACTION_PREFIX = '[ext]';
		const createExternalAction = createActionFactory(exports.EXTERNAL_ACTION_PREFIX);
		exports.iconsLoaded = createExternalAction('icons-loaded');
		exports.controlSelected = createExternalAction('control-selected');
		exports.selectControl = createExternalAction('select-control');
		exports.addExtensionPoint = createExternalAction('add-extension-point');
		exports.deletePropertyChanges = createExternalAction('delete-property-changes');
		exports.outlineChanged = createExternalAction('outline-changed');
		exports.changeProperty = createExternalAction('change-property');
		exports.propertyChanged = createExternalAction('property-changed');
		exports.propertyChangeFailed = createExternalAction('change-property-failed');
		exports.changeStackModified = createExternalAction('change-stack-modified');
		exports.showMessage = createExternalAction('show-dialog-message');
		exports.reloadApplication = createExternalAction('reload-application');
		exports.storageFileChanged = createExternalAction('storage-file-changed');
		exports.setAppMode = createExternalAction('set-app-mode');
		exports.applicationModeChanged = createExternalAction('application-mode-changed');
		exports.setUndoRedoEnablement = createExternalAction('set-undo-redo-enablement');
		exports.setSaveEnablement = createExternalAction('set-save-enablement');
		exports.appLoaded = createExternalAction('app-loaded');
		exports.undo = createExternalAction('undo');
		exports.redo = createExternalAction('redo');
		exports.save = createExternalAction('save');
		exports.quickActionListChanged = createExternalAction('quick-action-list-changed');
		exports.updateQuickAction = createExternalAction('update-quick-action');
		exports.executeQuickAction = createExternalAction('execute-quick-action');
		exports.executeContextMenuAction = createExternalAction('execute-context-menu-action');
		exports.setApplicationRequiresReload = createExternalAction('set-application-requires-reload');
		exports.externalFileChange = createExternalAction('external-file-change');
		exports.showInfoCenterMessage = createExternalAction('show-info-center-message');
		const createAsyncExternalAction = createAsyncActionFactory(exports.EXTERNAL_ACTION_PREFIX);
		exports.requestControlContextMenu = createAsyncExternalAction('request-control-context-menu');
		
	} (api));

	var debounce$1 = {};

	Object.defineProperty(debounce$1, "__esModule", { value: true });
	debounce$1.debounce = debounce;
	/**
	 * Returns a function which calls the callback function only after the specified idle time
	 * Works similar to the Debounce operator from rxjs https://reactivex.io/documentation/operators/debounce.html link.
	 *
	 * @param callback Function to execute
	 * @param delay Idle period in milliseconds after which the callback will be executed
	 * @returns A wrapper function that should be called to invoke the callback function after delay
	 */
	function debounce(callback, delay) {
	    let timerId;
	    return (...args) => {
	        clearTimeout(timerId);
	        timerId = setTimeout(() => {
	            callback(...args);
	        }, delay);
	    };
	}

	var postMessage = {};

	Object.defineProperty(postMessage, "__esModule", { value: true });
	postMessage.startPostMessageCommunication = startPostMessageCommunication;
	const POST_MESSAGE_ACTION_TYPE = 'post-message-action';
	/**
	 * Check if data isPostMessageAction.
	 *
	 * @param data - post message action
	 * @returns data is PostMessageAction<T>
	 */
	function isPostMessageAction(data) {
	    return data?.type === POST_MESSAGE_ACTION_TYPE && typeof data?.action === 'object';
	}
	/**
	 * Method to start post message communication.
	 *
	 * @param target target window
	 * @param onActionHandler action handler
	 * @param logger to log info, default: browser console.
	 * @returns PostMessageCommunication<T>
	 */
	function startPostMessageCommunication(target, onActionHandler, logger = console) {
	    /**
	     * Returns target windows or undefined.
	     *
	     * @returns Window | undefined
	     */
	    function getTarget() {
	        if (typeof target === 'function') {
	            return target();
	        }
	        return target;
	    }
	    /**
	     * Invoke action on post message.
	     *
	     * @param event event
	     */
	    function postMessageListener(event) {
	        const target = getTarget();
	        if (!target || event.origin !== target.origin || event.source !== target) {
	            // Ignore messages from unknown sources
	            return;
	        }
	        if (isPostMessageAction(event.data)) {
	            onActionHandler(event.data.action).catch((error) => logger.error(error));
	        }
	        else {
	            logger.warn(`Unknown message received ${event.data}`);
	        }
	    }
	    function dispose() {
	        window.removeEventListener('message', postMessageListener);
	    }
	    /**
	     * Post message to a give window.
	     *
	     * @param action action payload
	     */
	    function sendAction(action) {
	        const target = getTarget();
	        if (!target) {
	            return;
	        }
	        const message = {
	            type: POST_MESSAGE_ACTION_TYPE,
	            action
	        };
	        target.postMessage(message, target.origin);
	    }
	    window.addEventListener('message', postMessageListener);
	    return {
	        dispose,
	        sendAction
	    };
	}

	var telemetry = {};

	Object.defineProperty(telemetry, "__esModule", { value: true });
	telemetry.enableTelemetry = enableTelemetry;
	telemetry.disableTelemetry = disableTelemetry;
	telemetry.reportTelemetry = reportTelemetry;
	let enabled = false;
	function enableTelemetry() {
	    enabled = true;
	}
	function disableTelemetry() {
	    enabled = false;
	}
	/**
	 * Reports telemetry data from Control Property Editor.
	 *
	 * @param data The TelemetryData object, that needs to be reported
	 * @returns {Promise<void>}
	 */
	async function reportTelemetry(data) {
	    try {
	        if (enabled) {
	            const requestOptions = {
	                method: 'POST',
	                headers: {
	                    Accept: 'application/json',
	                    'Content-Type': 'application/json'
	                },
	                body: JSON.stringify(data)
	            };
	            await fetch('/preview/api/telemetry', requestOptions);
	        }
	    }
	    catch (_error) {
	        // something is wrong with the telemetry service
	        disableTelemetry();
	    }
	}

	var utils = {};

	Object.defineProperty(utils, "__esModule", { value: true });
	utils.FlexChangesEndPoints = utils.convertCamelCaseToPascalCase = void 0;
	const isUpperCase = (code) => code >= 65 && code <= 90;
	const isLowerCase = (code) => code >= 97 && code <= 122;
	const isWordDelimiter = (code) => code === 45 || code === 95 || code === 32; // - or _ or space
	const convertCamelCaseToPascalCase = (text) => {
	    const words = [];
	    let word = '';
	    let lookForUpperCase = true;
	    for (let i = 0; i < (text ?? '').length; i++) {
	        const character = text[i];
	        if (lookForUpperCase) {
	            // make sure that the first letter is capitalized
	            word += word.length === 0 ? character.toUpperCase() : character;
	            if (isLowerCase(text.charCodeAt(i + 1))) {
	                // First lower case character after upper case character -> switch mode to collect only lower case characters
	                lookForUpperCase = false;
	            }
	            else if (isUpperCase(text.charCodeAt(i + 1)) && isLowerCase(text.charCodeAt(i + 2))) {
	                // Next character is the last uppercase character after a sequence of upper case character -> create an abbreviated word
	                words.push(word);
	                word = '';
	            }
	        }
	        else if (isUpperCase(text.charCodeAt(i))) {
	            // Upper case character indicates the beginning of a new word -> switch mode to detect abbreviated word
	            words.push(word);
	            lookForUpperCase = true;
	            word = character;
	        }
	        else if (isWordDelimiter(text.charCodeAt(i))) {
	            words.push(word);
	            lookForUpperCase = true;
	            word = '';
	        }
	        else {
	            word += character;
	        }
	    }
	    if (word.length) {
	        words.push(word);
	    }
	    return words.join(' ');
	};
	utils.convertCamelCaseToPascalCase = convertCamelCaseToPascalCase;
	var FlexChangesEndPoints;
	(function (FlexChangesEndPoints) {
	    FlexChangesEndPoints["changes"] = "/preview/api/changes";
	})(FlexChangesEndPoints || (utils.FlexChangesEndPoints = FlexChangesEndPoints = {}));

	(function (exports) {
		var __createBinding = (commonjsGlobal && commonjsGlobal.__createBinding) || (Object.create ? (function(o, m, k, k2) {
		    if (k2 === undefined) k2 = k;
		    var desc = Object.getOwnPropertyDescriptor(m, k);
		    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
		      desc = { enumerable: true, get: function() { return m[k]; } };
		    }
		    Object.defineProperty(o, k2, desc);
		}) : (function(o, m, k, k2) {
		    if (k2 === undefined) k2 = k;
		    o[k2] = m[k];
		}));
		var __exportStar = (commonjsGlobal && commonjsGlobal.__exportStar) || function(m, exports) {
		    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
		};
		Object.defineProperty(exports, "__esModule", { value: true });
		exports.disableTelemetry = exports.enableTelemetry = exports.reportTelemetry = exports.startPostMessageCommunication = exports.debounce = void 0;
		__exportStar(api, exports);
		var debounce_1 = debounce$1;
		Object.defineProperty(exports, "debounce", { enumerable: true, get: function () { return debounce_1.debounce; } });
		var post_message_1 = postMessage;
		Object.defineProperty(exports, "startPostMessageCommunication", { enumerable: true, get: function () { return post_message_1.startPostMessageCommunication; } });
		var telemetry_1 = telemetry;
		Object.defineProperty(exports, "reportTelemetry", { enumerable: true, get: function () { return telemetry_1.reportTelemetry; } });
		Object.defineProperty(exports, "enableTelemetry", { enumerable: true, get: function () { return telemetry_1.enableTelemetry; } });
		Object.defineProperty(exports, "disableTelemetry", { enumerable: true, get: function () { return telemetry_1.disableTelemetry; } });
		__exportStar(utils, exports);
		
	} (dist));

	Object.defineProperty(dist, "__" + "esModule", { value: true });

	return dist;

}));
