// @sap-ux/ui5-info/dist/ui5-version-fallback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedUi5VersionFallbacks = exports.defaultUi5Versions = exports.ui5VersionFallbacks = exports.supportState = void 0;
const semver_1 = require("semver");
const constants_1 = require("./constants");
exports.supportState = {
    maintenance: 'Maintenance',
    outOfMaintenance: 'Out of maintenance',
    skipped: 'Skipped'
};
// Updated June-09-2025 from https://ui5.sap.com/versionoverview.json
exports.ui5VersionFallbacks = [
    {
        version: '1.136.*',
        support: exports.supportState.maintenance
    },
    {
        version: '1.135.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.134.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.133.*',
        support: exports.supportState.maintenance
    },
    {
        version: '1.132.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.131.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.130.*',
        support: exports.supportState.maintenance
    },
    {
        version: '1.129.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.128.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.127.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.126.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.125.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.124.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.123.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.122.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.121.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.120.*',
        support: exports.supportState.maintenance
    },
    {
        version: '1.119.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.118.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.117.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.116.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.115.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.114.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.113.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.112.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.111.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.110.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.109.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.108.*',
        support: exports.supportState.maintenance
    },
    {
        version: '1.107.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.106.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.105.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.104.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.103.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.102.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.101.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.100.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.99.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.98.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.97.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.96.*',
        support: exports.supportState.maintenance
    },
    {
        version: '1.95.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.94.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.93.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.92.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.91.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.90.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.89.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.88.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.87.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.86.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.85.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.84.*',
        support: exports.supportState.maintenance
    },
    {
        version: '1.83.*',
        support: exports.supportState.skipped
    },
    {
        version: '1.82.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.81.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.80.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.79.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.78.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.77.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.76.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.75.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.74.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.73.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.72.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.71.*',
        support: exports.supportState.maintenance
    },
    {
        version: '1.70.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.69.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.68.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.67.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.66.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.65.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.64.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.63.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.62.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.61.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.60.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.58.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.56.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.54.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.52.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.50.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.48.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.46.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.44.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.42.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.40.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.38.*',
        support: exports.supportState.maintenance
    },
    {
        version: '1.36.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.34.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.32.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.30.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.28.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.26.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.24.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '1.22.*',
        support: exports.supportState.outOfMaintenance
    },
    {
        version: '*',
        support: exports.supportState.outOfMaintenance
    }
];
const supportedUi5VersionFallbacks = exports.ui5VersionFallbacks
    .filter((supportVersion) => supportVersion.support === exports.supportState.maintenance &&
    (0, semver_1.gte)((0, semver_1.coerce)(supportVersion.version) ?? '0.0.0', constants_1.defaultMinUi5Version))
    .map((maintainedVersion) => ({
    version: (0, semver_1.coerce)(maintainedVersion.version)?.version ?? '0.0.0',
    support: exports.supportState.maintenance
}));
exports.supportedUi5VersionFallbacks = supportedUi5VersionFallbacks;
const defaultUi5Versions = [...supportedUi5VersionFallbacks].map((version) => version.version);
exports.defaultUi5Versions = defaultUi5Versions;
