// eslint-plugin-fiori-custom/lib/rules/sap-ui5-legacy-factories.js

// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
/* eslint-disable strict */
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
module.exports = {
    create: function(context) {
        "use strict";
        var MSG = "Make use of sap.ui.define([...], function(...) {...} to load required dependencies. Legacy UI5 factories leading to synchronous loading.";
        var INTERESTING_METHODS = ["jsview", "component", "view", "xmlview", "controller", "extensionpoint", "fragment", "getVersionInfo", "resources", "load"];
        var INTERESTING_METHODS_JQUERY = ["resources"];

        // --------------------------------------------------------------------------
        // Basic Helpers
        // --------------------------------------------------------------------------
        function isType(node, type) {
            return node && node.type === type;
        }

        function isIdentifier(node) {
            return isType(node, "Identifier");
        }

        function isLiteral(node) {
            return isType(node, "Literal");
        }

        function isProperty(node) {
            return isType(node, "Property");
        }

        function isMember(node) {
            return isType(node, "MemberExpression");
        }

        function isObject(node) {
            return isType(node, "ObjectExpression");
        }

        function contains(a, obj) {
            for (var i = 0; i < a.length; i++) {
                if (obj === a[i]) {
                    return true;
                }
            }
            return false;
        }

        function isNumber(i) {
            return Number(i) === i && i % 1 === 0;
        }

        function isInteresting(node) {
            var callee = node.callee;
            if (isMember(callee)) {
                if (isIdentifier(callee.property)  && contains(INTERESTING_METHODS, callee.property.name) ) {
                    if(isMember( callee.object) && callee.object.property.name=="ui")
                        if( isIdentifier(callee.object.object) && callee.object.object.name=="sap" )
                            return true;
                }
            }
            if (isMember(callee)) {
                if (isIdentifier(callee.property)  && contains(INTERESTING_METHODS_JQUERY, callee.property.name) ) {
                    if(isMember( callee.object) && callee.object.property.name=="sap")
                        if( isIdentifier(callee.object.object) && (callee.object.object.name=="jQuery" || callee.object.object.name=="$"))
                            return true;
                }
            }
            if (isMember(callee)) {
                if (isIdentifier(callee.property)  && contains(INTERESTING_METHODS, callee.property.name) ) {
                    if(isMember( callee.object) && callee.object.property.name=="component")
                        if( isMember(callee.object.object) && callee.object.object.property.name=="ui")
                            if( isIdentifier(callee.object.object.object) && callee.object.object.object.name=="sap")
                                return true;
                }
            }
            return false;
        }
  


        return {
            "CallExpression": function(node) {
                if (isInteresting(node)) {
                    context.report(node, MSG);
                }
            }
        };
    }
};
