// eslint-plugin-fiori-custom/lib/rules/sap-ui5-legacy-jquerysap-usage.js
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
/* eslint-disable strict */
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
module.exports = {
    create: function(context) {
        "use strict";
        var MSG = "Legacy jQuery.sap usage is not allowed due to strict Content Security Policy.";
      
        var INTERESTING_METHODS_JQUERY = ["require", "declare"];

        // --------------------------------------------------------------------------
        // Basic Helpers
        // --------------------------------------------------------------------------
        function isType(node, type) {
            return node && node.type === type;
        }

        function isIdentifier(node) {
            return isType(node, "Identifier");
        }

        function isMember(node) {
            return isType(node, "MemberExpression");
        }

        function contains(a, obj) {
            for (var i = 0; i < a.length; i++) {
                if (obj === a[i]) {
                    return true;
                }
            }
            return false;
        }

       

        function isInteresting(node) {
            var callee = node.callee;
            if (isMember(callee)) {
                if (isIdentifier(callee.property)  && contains(INTERESTING_METHODS_JQUERY, callee.property.name) ) {
                    if(isMember( callee.object) && callee.object.property.name=="sap")
                        if( isIdentifier(callee.object.object) && (callee.object.object.name=="jQuery" || callee.object.object.name=="$"))
                            return true;
                }
            }
            return false;
        } 
      


        return {
            "CallExpression": function(node) {
                if (isInteresting(node)) {
                    context.report(node, MSG);
                }
                
            }
        };
    }
};
