// @sap-ux/preview-middleware/dist/base/utils/cards.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIntegrationCard = getIntegrationCard;
const package_json_1 = __importDefault(require("../../../package.json"));
const CARD_TYPES = {
    INTEGRATION: 'integration',
    ADAPTIVE: 'adaptive'
};
/**
 * Prepares an integration card manifest for saving by adding the design time preview middleware i.e.`dtpMiddleware` version to the card manifest.
 *
 * @param {CardManifest} cardManifest - The card manifest object to be updated.
 * @returns {void}
 */
function prepareIntegrationCardForSaving(cardManifest) {
    const version = package_json_1.default.version;
    const insights = cardManifest?.['sap.insights'];
    if (!insights.versions) {
        insights.versions = {
            dtpMiddleware: version
        };
    }
    else {
        insights.versions.dtpMiddleware = version;
    }
}
/**
 * Retrieves the integration card from the provided payload and prepares it for saving.
 *
 * @param {MultiCardsPayload[]} multipleCard - The payload containing multiple card manifests.
 * @returns {MultiCardsPayload} The integration card manifest.
 */
function getIntegrationCard(multipleCard) {
    const integrationCard = multipleCard.find((card) => card.type === CARD_TYPES.INTEGRATION) ?? {
        type: CARD_TYPES.INTEGRATION,
        manifest: {
            'sap.insights': {}
        },
        entitySet: ''
    };
    prepareIntegrationCardForSaving(integrationCard.manifest);
    return integrationCard;
}
