// @sap-ux/adp-tooling/dist/writer/changes/writers/data-source-writer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataSourceWriter = void 0;
const change_utils_1 = require("../../../base/change-utils");
/**
 * Handles the creation and writing of data source data changes for a project.
 */
class DataSourceWriter {
    fs;
    projectPath;
    /**
     * @param {Editor} fs - The filesystem editor instance.
     * @param {string} projectPath - The root path of the project.
     */
    constructor(fs, projectPath) {
        this.fs = fs;
        this.projectPath = projectPath;
    }
    /**
     * Constructs content for a data source change.
     *
     * @param {string} dataSourceId - The ID of the data source being modified.
     * @param {string} dataSourceUri - The new URI to update the data source with.
     * @param {number} [maxAge] - Optional maximum age.
     * @returns {object} The constructed content object for the change data source change.
     */
    constructContent(dataSourceId, dataSourceUri, maxAge) {
        const content = {
            dataSourceId: dataSourceId,
            entityPropertyChange: [
                {
                    propertyPath: 'uri',
                    operation: 'UPDATE',
                    propertyValue: dataSourceUri
                }
            ]
        };
        if (maxAge) {
            content.entityPropertyChange.push({
                propertyPath: 'settings/maxAge',
                operation: 'UPSERT',
                propertyValue: Number(maxAge)
            });
        }
        return content;
    }
    /**
     * Writes the change data source change to the project based on the provided data.
     *
     * @param {DataSourceData} data - The change data source data containing all the necessary information to construct and write the change.
     * @returns {Promise<void>} A promise that resolves when the change writing process is completed.
     */
    async write(data) {
        const { variant, dataSources, service } = data;
        const { id, uri, maxAge, annotationUri } = service;
        const annotationId = dataSources[id].settings?.annotations?.[0];
        const timestamp = Date.now();
        const content = this.constructContent(id, uri, maxAge);
        const change = (0, change_utils_1.getChange)(variant, timestamp, content, "appdescr_app_changeDataSource" /* ChangeType.CHANGE_DATA_SOURCE */);
        (0, change_utils_1.writeChangeToFolder)(this.projectPath, change, this.fs);
        if (annotationId && annotationUri) {
            const annotationContent = this.constructContent(annotationId, annotationUri);
            const annotationTs = timestamp + 1;
            const annotationChange = (0, change_utils_1.getChange)(variant, annotationTs, annotationContent, "appdescr_app_changeDataSource" /* ChangeType.CHANGE_DATA_SOURCE */);
            (0, change_utils_1.writeChangeToFolder)(this.projectPath, annotationChange, this.fs);
        }
    }
}
exports.DataSourceWriter = DataSourceWriter;
