// @sap-ux/axios-extension/dist/abap/adt-catalog/services/abapcdsview-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbapCDSViewService = void 0;
const adt_service_1 = require("./adt-service");
/**
 * AbapCDSViewService implements ADT requests to obtain the list of abap cds views.
 */
class AbapCDSViewService extends adt_service_1.AdtService {
    /**
     * @see AdtService.getAdtCatagory()
     */
    static adtCategory = {
        scheme: 'http://www.sap.com/adt/categories/respository',
        term: 'search'
    };
    /**
     * Get ADT scheme ID.
     *
     * @returns AdtCategory
     */
    static getAdtCatagory() {
        return AbapCDSViewService.adtCategory;
    }
    /**
     * Get all abap cds views from ADT service.
     *
     * @param maxResults - The maximum number of abap cds views to be returned.
     * @returns A list of abap cds views.
     */
    async getAbapCDSViews(maxResults = 10000) {
        const config = {
            headers: {
                Accept: 'application/xml'
            },
            params: {
                operation: 'quickSearch',
                query: `*`,
                maxResults: maxResults,
                objectType: 'DDLS',
                releaseState: 'USE_IN_CLOUD_DEVELOPMENT'
            }
        };
        const response = await this.get('', config);
        return this.parseAbapCDSViewResponse(response.data);
    }
    /**
     * Parse the XML document of abap cds view entries from ADT service.
     *
     * @param xml xml document containing abap cds view entries.
     * @returns A list of abap cds view names.
     */
    parseAbapCDSViewResponse(xml) {
        const parsed = this.parseResponse(xml);
        let abapCDSViewArray = [];
        if (parsed?.objectReferences?.objectReference) {
            if (Array.isArray(parsed.objectReferences.objectReference)) {
                abapCDSViewArray = parsed.objectReferences.objectReference;
            }
            else {
                abapCDSViewArray = [parsed.objectReferences.objectReference];
            }
        }
        return abapCDSViewArray.map((item) => {
            const { name, uri, description } = item;
            return { name, uri, description };
        });
    }
}
exports.AbapCDSViewService = AbapCDSViewService;
