// @sap-ux/axios-extension/dist/abap/adt-catalog/services/adt-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdtService = void 0;
const axios_1 = require("axios");
const fast_xml_parser_1 = require("fast-xml-parser");
/**
 * AdtService provides abstraction of AdtService implementations. Each specific ADT service
 * should be implemented as subclass of AdtService.
 */
class AdtService extends axios_1.Axios {
    // Instantiated by calling ServiceProvider.createService()
    log;
    // ADT schema for the corresponding AdtService subclass
    serviceSchema;
    /**
     * Subclass that implements each specific ADT service
     * should provide AdtCatagory for the corresponding AdtService.
     * The AdtCatagory properties combined to identify a particular
     * ADT service schema in ADT schema.
     */
    static getAdtCatagory() {
        throw new Error('NEEDS TO BE IMPLEMENTED BY AdtService SUBCLASSES');
    }
    /**
     * Attach schema to AdtService subclass.
     *
     * @param serviceSchema ADT schema for this particular ADT service catalog.
     */
    attachAdtSchema(serviceSchema) {
        this.serviceSchema = serviceSchema;
    }
    /**
     * Parse an XML document for ATO (Adaptation Transport Organizer) settings.
     *
     * @param xml xml document containing ATO settings
     * @returns parsed ATO settings
     */
    parseResponse(xml) {
        if (fast_xml_parser_1.XMLValidator.validate(xml) !== true) {
            this.log.warn(`Invalid XML: ${xml}`);
            return {};
        }
        const options = {
            attributeNamePrefix: '',
            ignoreAttributes: false,
            ignoreNameSpace: true,
            parseAttributeValue: true,
            removeNSPrefix: true
        };
        const parser = new fast_xml_parser_1.XMLParser(options);
        return parser.parse(xml, true);
    }
}
exports.AdtService = AdtService;
