// @sap-ux/axios-extension/dist/abap/adt-catalog/services/generator-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeneratorService = void 0;
const adt_service_1 = require("./adt-service");
/**
 *
 */
class GeneratorService extends adt_service_1.AdtService {
    /**
     * @see AdtService.getAdtCatagory()
     */
    static adtCategory = {
        scheme: 'http://www.sap.com/adt/categories/respository',
        term: 'generators'
    };
    /**
     * Get ADT scheme ID.
     *
     * @returns AdtCategory
     */
    static getAdtCatagory() {
        return GeneratorService.adtCategory;
    }
    id;
    /**
     * Get the UI service generator for the given referenced object.
     *
     * @param objectUri - The object (business object or abap cds view) uri.
     * @returns TBD
     */
    async getUIServiceGeneratorConfig(objectUri) {
        const response = await this.get('', {
            headers: {
                Accept: 'application/atom+xml;type=feed'
            },
            params: {
                referencedObject: objectUri,
                type: 'webapi'
            }
        });
        const data = this.parseResponse(response.data).feed?.entry;
        if (data?.id === 'published-ui-service') {
            return data;
        }
        else {
            throw new Error('UI Service Generator not found');
        }
    }
}
exports.GeneratorService = GeneratorService;
