// @sap-ux/axios-extension/dist/abap/adt-catalog/services/publish-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublishService = void 0;
const adt_service_1 = require("./adt-service");
/**
 *
 */
class PublishService extends adt_service_1.AdtService {
    /**
     * @see AdtService.getAdtCatagory()
     */
    static adtCategory = {
        scheme: 'http://www.sap.com/categories/servicebindings/bindingtypes',
        term: 'ODataV4'
    };
    /**
     * Get ADT scheme ID.
     *
     * @returns AdtCategory
     */
    static getAdtCatagory() {
        return PublishService.adtCategory;
    }
    /**
     * Get OData V4 service URI for the given business object.
     *
     * @param technicalDetails - technical name of OData service
     * @returns service URI.
     */
    async getODataV4ServiceUri(technicalDetails) {
        const { serviceDefinitionName, serviceName, serviceVersion } = technicalDetails;
        const response = await this.get(`/${serviceName}`, {
            headers: {
                Accept: 'application/vnd.sap.adt.businessservices.odatav4.v2+xml'
            },
            params: {
                servicename: serviceName,
                serviceversion: serviceVersion,
                srvdname: serviceDefinitionName
            }
        });
        const data = this.parseResponse(response.data);
        return String(data.serviceGroup.services.serviceUrl);
    }
}
exports.PublishService = PublishService;
