// @sap-ux/axios-extension/dist/abap/adt-catalog/services/rap-generator-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RapGeneratorService = void 0;
const adt_service_1 = require("./adt-service");
/**
 * RapGeneratorService implements ADT requests for generating an OData Service
 * from ABAP backend system.
 */
class RapGeneratorService extends adt_service_1.AdtService {
    /**
     * @see AdtService.getAdtCategory()
     */
    static adtCategory = {
        scheme: 'http://www.sap.com/adt/categories/respository',
        term: 'generators'
    };
    /**
     * Get ADT scheme ID.
     *
     * @returns AdtCategory
     */
    static getAdtCatagory() {
        return RapGeneratorService.adtCategory;
    }
    id;
    /**
     * Get the RAP Generator (published-x-ui-service).
     *
     * @returns Promise<GeneratorEntry>
     */
    async getRAPGeneratorConfig() {
        const response = await this.get('', {
            headers: {
                Accept: 'application/atom+xml;type=feed'
            },
            params: {
                type: 'webapi',
                fetchAllGenerators: true
            }
        });
        let generators = (this.parseResponse(response.data).feed?.entry ?? []);
        if (generators && !Array.isArray(generators)) {
            generators = [generators];
        }
        const data = generators.find((generator) => generator.id === 'published-x-ui-service');
        if (data) {
            return data;
        }
        else {
            throw new Error('OData Service Generator not found');
        }
    }
}
exports.RapGeneratorService = RapGeneratorService;
