// @sap-ux/axios-extension/dist/abap/adt-catalog/services/transportrequest-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransportRequestService = void 0;
const adt_service_1 = require("./adt-service");
/**
 * TransportRequestService implements ADT requests for creating a new transport request number
 * that can be used for deploying Fiori project to ABAP backend.
 */
class TransportRequestService extends adt_service_1.AdtService {
    /**
     * @see AdtService.getAdtCatagory()
     */
    static adtCategory = {
        scheme: 'http://www.sap.com/adt/categories/cts',
        term: 'transports'
    };
    /**
     * @see AdtService.getAdtCatagory()
     * @returns AdtCategory
     */
    static getAdtCatagory() {
        return TransportRequestService.adtCategory;
    }
    /**
     * TransportRequestService API function to create a new transport number.
     *
     * @param reqParam Request parameter for creating a new transport request for an UI5 app object.
     * @returns Newly created transport request number
     */
    async createTransportRequest(reqParam) {
        const acceptHeaders = {
            headers: {
                Accept: 'text/plain',
                'content-type': 'application/vnd.sap.as+xml; charset=UTF-8; dataname=com.sap.adt.CreateCorrectionRequest'
            }
        };
        const data = `
                <?xml version="1.0" encoding="UTF-8"?>
                <asx:abap xmlns:asx="http://www.sap.com/abapxml" version="1.0">
                    <asx:values>
                        <DATA>
                            <OPERATION>I</OPERATION>
                            <DEVCLASS>${reqParam.packageName}</DEVCLASS>
                            <REQUEST_TEXT>${reqParam.description}</REQUEST_TEXT>
                            <REF>/sap/bc/adt/filestore/ui5-bsp/objects/${reqParam.ui5AppName}/$create</REF>
                        </DATA>
                    </asx:values>
                </asx:abap>
            `;
        const response = await this.post('', data, acceptHeaders);
        return this.getTransportNumberFromResponse(response.data);
    }
    /**
     * Read the newly created transport number from response XML data.
     *
     * @param text XML response of create transport request.
     * @returns Newly created transport number or null
     */
    getTransportNumberFromResponse(text) {
        const responseStringPrefix = '/com.sap.cts/object_record/';
        if (!text?.startsWith(responseStringPrefix)) {
            return null;
        }
        else {
            const newTransportNumber = text.replace(responseStringPrefix, '');
            return newTransportNumber;
        }
    }
}
exports.TransportRequestService = TransportRequestService;
