// @sap-ux/adp-tooling/dist/base/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ABAP_VARIANT_APPS_PARAMS = exports.ABAP_APPS_PARAMS = exports.S4HANA_APPS_PARAMS = exports.BASE_I18N_DESCRIPTION = exports.MAIN_I18N_PATH = exports.TRANSLATION_UUID_TEXT = exports.RESOURCE_BUNDLE_TEXT = exports.SNAPSHOT_UNTESTED_VERSION = exports.SNAPSHOT_VERSION = exports.LATEST_VERSION = exports.CURRENT_SYSTEM_VERSION = exports.SNAPSHOT_CDN_URL = exports.UI5_VERSIONS_NEO_CDN_URL = exports.UI5_VERSIONS_CDN_URL = exports.UI5_CDN_URL = void 0;
exports.UI5_CDN_URL = 'https://ui5.sap.com';
exports.UI5_VERSIONS_CDN_URL = 'https://sapui5.hana.ondemand.com/version.json';
exports.UI5_VERSIONS_NEO_CDN_URL = 'https://ui5.sap.com/neo-app.json';
exports.SNAPSHOT_CDN_URL = 'https://sapui5preview-sapui5.dispatcher.int.sap.eu2.hana.ondemand.com:443';
exports.CURRENT_SYSTEM_VERSION = '(system version)';
exports.LATEST_VERSION = '(latest)';
exports.SNAPSHOT_VERSION = 'snapshot';
exports.SNAPSHOT_UNTESTED_VERSION = 'snapshot-untested';
exports.RESOURCE_BUNDLE_TEXT = '\n\n# This is the resource bundle for ';
exports.TRANSLATION_UUID_TEXT = '\n#__ldi.translation.uuid=';
exports.MAIN_I18N_PATH = 'i18n/i18n.properties';
exports.BASE_I18N_DESCRIPTION = '#Make sure you provide a unique prefix to the newly added keys in this file, to avoid overriding of SAP Fiori application keys.';
exports.S4HANA_APPS_PARAMS = {
    'sap.app/type': 'application',
    'sap.fiori/cloudDevAdaptationStatus': 'released',
    'fields': 'sap.app/id,repoName,sap.fiori/cloudDevAdaptationStatus,sap.app/ach,sap.fiori/registrationIds,sap.app/title,url,fileType'
};
exports.ABAP_APPS_PARAMS = {
    'fields': 'sap.app/id,sap.app/ach,sap.fiori/registrationIds,sap.app/title,url,fileType,repoName',
    'sap.ui/technology': 'UI5',
    'sap.app/type': 'application',
    'fileType': 'appdescr'
};
exports.ABAP_VARIANT_APPS_PARAMS = {
    'fields': 'sap.app/id,sap.app/ach,sap.fiori/registrationIds,sap.app/title,url,fileType,repoName',
    'sap.ui/technology': 'UI5',
    'sap.app/type': 'application',
    'fileType': 'appdescr_variant',
    'originLayer': 'VENDOR'
};
