// @sap-ux/telemetry/dist/base/performance/entries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Measurement = exports.Mark = void 0;
const types_1 = require("./types");
/**
 * Repreeents a mark during a performance measurement.
 */
class Mark {
    startTime;
    name;
    type;
    /**
     * @returns start time in millieseconds
     */
    getStartTime() {
        return this.startTime;
    }
    /**
     *
     * @param name Name of mark
     * @param type Mark type
     * @param startTime start time in millieseconds
     */
    constructor(name, type, startTime) {
        this.startTime = startTime;
        this.name = name;
        this.startTime = startTime;
        this.type = type;
    }
}
exports.Mark = Mark;
/**
 * Measurement of execution time length
 */
class Measurement extends Mark {
    startTime;
    duration;
    /**
     *
     * @param name Name of a mark
     * @param startTime start time in millieseconds
     * @param duration time in millieseconds
     */
    constructor(name, startTime, duration) {
        super(name, types_1.EntryType.MEASUREMENT, startTime);
        this.startTime = startTime;
        this.duration = duration;
        this.duration = duration;
    }
    /**
     * @returns duration in milliseconds
     */
    getDurationTime() {
        return this.duration;
    }
}
exports.Measurement = Measurement;
