// wdio-ui5-service/cjs/lib/authentication/BTPAuthenticator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Authenticator_js_1 = __importDefault(require("./Authenticator.js"));
var BTPAuthenticator = /** @class */ (function (_super) {
    __extends(BTPAuthenticator, _super);
    function BTPAuthenticator(options, browserInstanceName) {
        var _a, _b, _c, _d, _e, _f;
        var _this = _super.call(this, browserInstanceName) || this;
        _this.usernameSelector = (_a = options.usernameSelector) !== null && _a !== void 0 ? _a : "#j_username";
        _this.passwordSelector = (_b = options.passwordSelector) !== null && _b !== void 0 ? _b : "#j_password";
        _this.submitSelector = (_c = options.submitSelector) !== null && _c !== void 0 ? _c : "#logOnFormSubmit";
        _this.idpDomainOpt = (_e = (_d = options.idpDomain) === null || _d === void 0 ? void 0 : _d.trim()) !== null && _e !== void 0 ? _e : "";
        _this.disableBiometricAuth = (_f = options.disableBiometricAuthentication) !== null && _f !== void 0 ? _f : false;
        if (_this.disableBiometricAuth) {
            if (!options.idpDomain) {
                throw new Error("idpDomain is required when disableBiometricAuthentication is set to `true` your wdi5.conf.(j|t)s");
            }
            _this.idpDomain = options.idpDomain;
        }
        return _this;
    }
    BTPAuthenticator.prototype.disableBiometricAuthentication = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.browserInstance.setCookies({
                            name: "skipPasswordlessAuthnDeviceConfig",
                            value: "true",
                            domain: this.idpDomain
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    BTPAuthenticator.prototype.login = function () {
        return __awaiter(this, void 0, void 0, function () {
            var targetIdpEle, usernameControl, passwordControl, submit, wdi5Username, wdi5Password, password;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getIsLoggedIn()];
                    case 1:
                        if (!!(_a.sent())) return [3 /*break*/, 18];
                        if (!this.idpDomainOpt) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.browserInstance.$("a[href*=\"idp=".concat(this.idpDomainOpt, "\"]"))];
                    case 2:
                        targetIdpEle = _a.sent();
                        if (!!targetIdpEle.elementId) {
                            targetIdpEle.click();
                        }
                        _a.label = 3;
                    case 3: return [4 /*yield*/, this.browserInstance.$(this.usernameSelector)];
                    case 4:
                        usernameControl = _a.sent();
                        return [4 /*yield*/, this.browserInstance.$(this.passwordSelector)];
                    case 5:
                        passwordControl = _a.sent();
                        return [4 /*yield*/, this.browserInstance.$(this.submitSelector)];
                    case 6:
                        submit = _a.sent();
                        wdi5Username = this.getUsername();
                        wdi5Password = this.getPassword();
                        return [4 /*yield*/, passwordControl.isExisting()];
                    case 7:
                        if (!_a.sent()) return [3 /*break*/, 11];
                        return [4 /*yield*/, usernameControl.setValue(wdi5Username)];
                    case 8:
                        _a.sent();
                        return [4 /*yield*/, passwordControl.setValue(wdi5Password)];
                    case 9:
                        _a.sent();
                        return [4 /*yield*/, submit.click()];
                    case 10:
                        _a.sent();
                        return [3 /*break*/, 17];
                    case 11: return [4 /*yield*/, usernameControl.setValue(wdi5Username)];
                    case 12:
                        _a.sent();
                        return [4 /*yield*/, submit.click()];
                    case 13:
                        _a.sent();
                        return [4 /*yield*/, this.browserInstance.$(this.passwordSelector)];
                    case 14:
                        password = _a.sent();
                        return [4 /*yield*/, password.setValue(wdi5Password)];
                    case 15:
                        _a.sent();
                        return [4 /*yield*/, submit.click()];
                    case 16:
                        _a.sent();
                        _a.label = 17;
                    case 17:
                        this.setIsLoggedIn(true);
                        _a.label = 18;
                    case 18: return [2 /*return*/];
                }
            });
        });
    };
    return BTPAuthenticator;
}(Authenticator_js_1.default));
exports.default = BTPAuthenticator;
