// @sap-ux/preview-middleware/dist/client/adp/quick-actions/fe-v4/op-add-custom-section.js
"use strict";

sap.ui.define(["../../../utils/fe-v4", "../simple-quick-action-base", "../common/op-add-custom-section", "../dialog-enablement-validator", "../../../cpe/quick-actions/utils", "sap/ui/dt/OverlayRegistry", "../../dialog-factory"], function (_____utils_fe_v4, ___simple_quick_action_base, ___common_op_add_custom_section, ___dialog_enablement_validator, _____cpe_quick_actions_utils, OverlayRegistry, ____dialog_factory) {
  "use strict";

  const getV4AppComponent = _____utils_fe_v4["getV4AppComponent"];
  const SimpleQuickActionDefinitionBase = ___simple_quick_action_base["SimpleQuickActionDefinitionBase"];
  const OP_ADD_CUSTOM_SECTION = ___common_op_add_custom_section["OP_ADD_CUSTOM_SECTION"];
  const DIALOG_ENABLEMENT_VALIDATOR = ___dialog_enablement_validator["DIALOG_ENABLEMENT_VALIDATOR"];
  const getRelevantControlFromActivePage = _____cpe_quick_actions_utils["getRelevantControlFromActivePage"];
  const DialogFactory = ____dialog_factory["DialogFactory"];
  const DialogNames = ____dialog_factory["DialogNames"];
  const OP_PAGE_LAYOUT = ['sap.uxap.ObjectPageLayout'];
  /**
   * Quick Action for adding a custom page action.
   */
  class AddCustomSectionQuickAction extends SimpleQuickActionDefinitionBase {
    get currentPageDescriptor() {
      const appComponent = getV4AppComponent(this.context.view);
      const projectId = this.context.flexSettings.projectId;
      const objectPageLayout = getRelevantControlFromActivePage(this.context.controlIndex, this.context.view, OP_PAGE_LAYOUT)[0];
      const sections = objectPageLayout.getSections();
      let anchor = null;
      const pageId = this.context.view.getViewData()?.stableId.split('::').pop();
      if (sections.length > 0) {
        // Use the first section as the anchor if available
        anchor = (this.context.view.getLocalId(sections[sections.length - 1].getId()) ?? '').split('::').pop();
      }
      if (!pageId) {
        throw new Error('pageId is not defined');
      }
      if (!projectId) {
        throw new Error('app reference is not defined');
      }
      if (!appComponent) {
        throw new Error('appComponent is not defined');
      }
      if (!anchor) {
        throw new Error('appComponent is not defined');
      }
      return {
        appType: 'fe-v4',
        appComponent,
        anchor,
        pageId,
        projectId
      };
    }
    constructor(context) {
      super(OP_ADD_CUSTOM_SECTION, OP_PAGE_LAYOUT, 'QUICK_ACTION_OP_ADD_CUSTOM_SECTION', context, [DIALOG_ENABLEMENT_VALIDATOR]);
    }
    async execute() {
      const objectPageLayout = getRelevantControlFromActivePage(this.context.controlIndex, this.context.view, OP_PAGE_LAYOUT)[0];
      const overlay = OverlayRegistry.getOverlay(objectPageLayout) || [];
      await DialogFactory.createDialog(overlay, this.context.rta, DialogNames.ADD_CUSTOM_FRAGMENT, undefined, {
        propertyPath: 'content/body/sections/',
        title: 'QUICK_ACTION_OP_ADD_CUSTOM_SECTION',
        appDescriptor: this.currentPageDescriptor
      }, {
        actionName: this.type,
        telemetryEventIdentifier: this.getTelemetryIdentifier()
      });
      return [];
    }
  }
  var __exports = {
    __esModule: true
  };
  __exports.AddCustomSectionQuickAction = AddCustomSectionQuickAction;
  return __exports;
});
