// @sap-ux/telemetry/dist/tooling-telemetry/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonProperties = exports.DeployTarget = exports.ODataSource = exports.ToolsId = void 0;
var ToolsId;
(function (ToolsId) {
    ToolsId["UNKNOWN"] = "UNKNOWN";
    ToolsId["NO_TOOLS_ID"] = "NO_TOOLS_ID";
})(ToolsId || (exports.ToolsId = ToolsId = {}));
var ODataSource;
(function (ODataSource) {
    ODataSource["CAPJava"] = "CAPJava";
    ODataSource["CAPNode"] = "CAPNode";
    ODataSource["ABAP"] = "ABAP";
    ODataSource["UNKNOWN"] = "UNKNOWN";
})(ODataSource || (exports.ODataSource = ODataSource = {}));
var DeployTarget;
(function (DeployTarget) {
    DeployTarget["CF"] = "CF";
    DeployTarget["ABAP"] = "ABAP";
    DeployTarget["UNKNOWN_DEPLOY_CONFIG"] = "UNKNOWN_DEPLOY_CONFIG";
    DeployTarget["NO_DEPLOY_CONFIG"] = "NO_DEPLOY_CONFIG";
})(DeployTarget || (exports.DeployTarget = DeployTarget = {}));
var CommonProperties;
(function (CommonProperties) {
    CommonProperties["DevSpace"] = "cmn.devspace";
    CommonProperties["AppStudio"] = "cmn.appstudio";
    CommonProperties["AppStudioBackwardCompatible"] = "appstudio";
    CommonProperties["InternlVsExternal"] = "cmn.internalFeatures";
    CommonProperties["InternlVsExternalBackwardCompatible"] = "internalVsExternal";
    CommonProperties["TemplateType"] = "cmn.template";
    CommonProperties["DeployTargetType"] = "cmn.deployTarget";
    CommonProperties["ODataSourceType"] = "cmn.odataSource";
    CommonProperties["AppToolsId"] = "cmn.toolsId";
    CommonProperties["NodeVersion"] = "cmn.nodeVersion";
    CommonProperties["AppProgrammingLanguage"] = "cmn.appLanguage";
    CommonProperties["TemplateId"] = "cmn.templateId";
    CommonProperties["TemplateVersion"] = "cmn.templateVersion";
    CommonProperties["ApplicationType"] = "cmn.applicationType";
})(CommonProperties || (exports.CommonProperties = CommonProperties = {}));
