// @sap-ux/adp-tooling/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateInboundConfig = exports.getCustomConfig = exports.generateChange = exports.migrate = exports.generate = exports.promptGeneratorInput = void 0;
__exportStar(require("./types"), exports);
__exportStar(require("./prompts"), exports);
__exportStar(require("./common"), exports);
__exportStar(require("./abap"), exports);
__exportStar(require("./source"), exports);
__exportStar(require("./ui5"), exports);
__exportStar(require("./base/cf"), exports);
__exportStar(require("./base/helper"), exports);
__exportStar(require("./base/constants"), exports);
__exportStar(require("./base/project-builder"), exports);
__exportStar(require("./base/abap/manifest-service"), exports);
var prompt_1 = require("./base/prompt");
Object.defineProperty(exports, "promptGeneratorInput", { enumerable: true, get: function () { return prompt_1.promptGeneratorInput; } });
__exportStar(require("./preview/adp-preview"), exports);
__exportStar(require("./writer/manifest"), exports);
__exportStar(require("./writer/writer-config"), exports);
var writer_1 = require("./writer");
Object.defineProperty(exports, "generate", { enumerable: true, get: function () { return writer_1.generate; } });
Object.defineProperty(exports, "migrate", { enumerable: true, get: function () { return writer_1.migrate; } });
var editors_1 = require("./writer/editors");
Object.defineProperty(exports, "generateChange", { enumerable: true, get: function () { return editors_1.generateChange; } });
var project_utils_1 = require("./writer/project-utils");
Object.defineProperty(exports, "getCustomConfig", { enumerable: true, get: function () { return project_utils_1.getCustomConfig; } });
var inbound_navigation_1 = require("./writer/inbound-navigation");
Object.defineProperty(exports, "generateInboundConfig", { enumerable: true, get: function () { return inbound_navigation_1.generateInboundConfig; } });
