// @sap/bas-sdk/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ai = exports.cfendpoint = exports.remotessh = exports.hrtt = exports.devspace = exports.destinations = exports.devspaceApi = exports.core = void 0;
const isAppStudio = __importStar(require("./apis/is-app-studio"));
const headersHelper = __importStar(require("./apis/headers-helper"));
const getDestinations = __importStar(require("./apis/get-destinations"));
const createDestination = __importStar(require("./apis/create-destination"));
const devspaceApiFactory = __importStar(require("./apis/devspace-api-factory"));
const hanaHrttHelper = __importStar(require("./apis/hrtt-helper"));
const getEnvValue = __importStar(require("./apis/get-env-value"));
const runCommand = __importStar(require("./apis/run-command"));
const getCFEndpoint = __importStar(require("./apis/get-cf-endpoint"));
const get_open_project_in_vscode_uri_1 = require("./apis/get-open-project-in-vscode-uri");
const core_utils_1 = require("./utils/core-utils");
const get_devspace_1 = require("./apis/get-devspace");
const didBASGeneratorsFinishInstallation = __importStar(require("./apis/get-generators-installation-status"));
const devspace_utils_1 = require("./utils/devspace-utils");
const isInternalAccount = __importStar(require("./apis/is-internal-account"));
const ai_1 = require("./apis/ai");
const jwt_helper_1 = require("./apis/jwt-helper");
exports.core = Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, isAppStudio), headersHelper), runCommand), getEnvValue), { getExtLoginPath: core_utils_1.getExtLoginPath }), isInternalAccount), { getOpenProjectInVSCodeURI: get_open_project_in_vscode_uri_1.getOpenProjectInVSCodeURI,
    isJwtExpired: jwt_helper_1.isJwtExpired,
    timeUntilJwtExpires: jwt_helper_1.timeUntilJwtExpires });
exports.devspaceApi = Object.assign({}, devspaceApiFactory);
exports.destinations = Object.assign(Object.assign({}, getDestinations), createDestination);
exports.devspace = Object.assign(Object.assign({ getDevspaceInfo: get_devspace_1.getDevspaceInfo,
    getDevSpaces: get_devspace_1.getDevSpaces,
    deleteDevSpace: get_devspace_1.deleteDevSpace,
    updateDevSpace: get_devspace_1.updateDevSpace,
    isBuildCode: get_devspace_1.isBuildCode,
    getTenantMetadata: get_devspace_1.getTenantMetadata,
    getBasMode: get_devspace_1.getBasMode }, didBASGeneratorsFinishInstallation), { DevSpaceStatus: devspace_utils_1.DevSpaceStatus,
    PackName: devspace_utils_1.PackName,
    getAppExternalUri: get_devspace_1.getAppExternalUri,
    calculateAppURLByPort: get_devspace_1.calculateAppURLByPort,
    isBuild: get_devspace_1.isBuild });
exports.hrtt = Object.assign({}, hanaHrttHelper);
exports.remotessh = {
    getKey: devspace_utils_1.getKey,
};
exports.cfendpoint = Object.assign({}, getCFEndpoint);
exports.ai = {
    sendAiRequest: ai_1.sendAiRequest,
};
