// @sap-ux/feature-toggle/dist/featureToggle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeatureToggleAccess = void 0;
exports.enableFeature = enableFeature;
exports.isFeatureEnabled = isFeatureEnabled;
exports.isInternalFeaturesSettingEnabled = isInternalFeaturesSettingEnabled;
const constants_1 = require("./constants");
/**
 * Utility class for accessing and managing feature toggles.
 */
class FeatureToggleAccess {
    static vscode = getVSCodeInstance();
    /**
     * Retrieves the toggle state of the specified feature.
     *
     * @param feature the feature to retrieve the toggle state for.
     * @returns the toggle state of the feature.
     */
    static getFeatureToggle(feature) {
        let toggleConfigValue;
        if ((feature.includes(constants_1.FeatureToggleKey) || feature === constants_1.ExperimentalFeatures) && FeatureToggleAccess.vscode) {
            const toggleKey = feature.slice(0, feature.lastIndexOf('.'));
            const toggleId = feature.slice(feature.lastIndexOf('.') + 1, feature.length);
            toggleConfigValue = FeatureToggleAccess.vscode.workspace.getConfiguration(toggleKey)?.get(toggleId);
        }
        else {
            toggleConfigValue = false;
        }
        // if TOOLSUITE_FEATURES env is set check if the feature is enabled there.
        if (process.env.TOOLSUITE_FEATURES) {
            const envFeatures = process.env.TOOLSUITE_FEATURES.split(',');
            toggleConfigValue = envFeatures.includes(feature) ? true : toggleConfigValue;
        }
        // toggle token matches so set to true
        if (constants_1.tokenToggleGuid[feature]) {
            if (constants_1.tokenToggleGuid[feature] === toggleConfigValue) {
                toggleConfigValue = true;
            }
            else {
                // it is a token toggle, but the token does not match
                toggleConfigValue = false;
            }
        }
        const featureToggle = {
            feature: feature,
            isEnabled: toggleConfigValue === true ? toggleConfigValue : false
        };
        return featureToggle;
    }
    /**
     * Retrieves all feature toggles.
     *
     * @returns an array of defined feature toggles.
     */
    static getAllFeatureToggles() {
        const definedToggles = [];
        if (FeatureToggleAccess.vscode) {
            Object.keys(constants_1.extensionConfigKeys).forEach((toggleConfigKey) => {
                const toggleKey = `${constants_1.extensionConfigKeys[toggleConfigKey]}.${constants_1.FeatureToggleKey}`;
                let toggles = {};
                try {
                    toggles = JSON.parse(JSON.stringify(FeatureToggleAccess.vscode.workspace.getConfiguration(toggleKey)));
                }
                catch {
                    // Not valid toggles. Skip.
                }
                Object.keys(toggles).forEach((toggleId) => {
                    // get full toggle value
                    const toggleConfigValue = FeatureToggleAccess.vscode.workspace
                        .getConfiguration(`${toggleKey}`)
                        .get(`${toggleId}`);
                    const toggle = {
                        feature: `${toggleKey}.${toggleId}`,
                        isEnabled: toggleConfigValue ? toggleConfigValue : false
                    };
                    definedToggles.push(toggle);
                });
            });
        }
        // if TOOLSUITE_FEATURES env is set check if the feature is enabled there.
        else if (process.env.TOOLSUITE_FEATURES) {
            const envFeatures = process.env.TOOLSUITE_FEATURES.split(',');
            for (const feature of envFeatures) {
                const toggle = {
                    feature,
                    isEnabled: true
                };
                definedToggles.push(toggle);
            }
        }
        return definedToggles;
    }
}
exports.FeatureToggleAccess = FeatureToggleAccess;
/**
 * Returns an instance of vscode vscode if available.
 *
 * @returns instance of vscode
 */
function getVSCodeInstance() {
    let vscode;
    try {
        vscode = require('vscode');
    }
    catch {
        // Vscode not available. Normally in CLI
    }
    return vscode;
}
/**
 * Enables a feature without a vscode reference by adding it to the TOOLSUITE_FEATURES environment variables.
 *
 * @param feature name of feature
 */
function enableFeature(feature) {
    let envFeatures = [];
    if (process.env.TOOLSUITE_FEATURES) {
        envFeatures = process.env.TOOLSUITE_FEATURES.split(',');
        if (!envFeatures.includes(feature)) {
            envFeatures.push(feature);
        }
    }
    else {
        envFeatures.push(feature);
    }
    process.env.TOOLSUITE_FEATURES = envFeatures.join();
}
/**
 * Checks if the provided feature toggle is enabled.
 *
 * @param feature name of feature
 * @returns {boolean} true if feature is enabled
 */
function isFeatureEnabled(feature) {
    return FeatureToggleAccess.getFeatureToggle(feature).isEnabled;
}
/**
 * Checks the internal enablement extension's `sap.ux.internal.enableInternalFeatures` config setting.
 *
 * @returns {boolean} true if the internal enablement is active
 */
function isInternalFeaturesSettingEnabled() {
    const enableInternalFeaturesSetting = 'sap.ux.internal.enableInternalFeatures';
    let internalEnabled = false;
    if (FeatureToggleAccess.vscode) {
        const internalSetting = FeatureToggleAccess.vscode.workspace
            ? FeatureToggleAccess.vscode.workspace.getConfiguration()?.get(enableInternalFeaturesSetting)
            : false;
        internalEnabled = internalSetting ?? false;
    }
    if (process.env.TOOLSUITE_INTERNAL && process.env.TOOLSUITE_INTERNAL === 'true') {
        internalEnabled = true;
    }
    return internalEnabled;
}
