// @sap-ux/fiori-generator-shared/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSemanticObject = exports.getFlpId = exports.getHostEnvironment = exports.generateAppGenInfo = exports.isExtensionInstalled = exports.getDefaultTargetFolder = exports.getBootstrapResourceUrls = exports.getPackageScripts = void 0;
__exportStar(require("./cap"), exports);
__exportStar(require("./constants"), exports);
__exportStar(require("./environment"), exports);
__exportStar(require("./system-utils"), exports);
__exportStar(require("./telemetry"), exports);
__exportStar(require("./logging/logWrapper"), exports);
__exportStar(require("./types"), exports);
var getPackageScripts_1 = require("./npm-package-scripts/getPackageScripts");
Object.defineProperty(exports, "getPackageScripts", { enumerable: true, get: function () { return getPackageScripts_1.getPackageScripts; } });
var ui5_1 = require("./ui5/ui5");
Object.defineProperty(exports, "getBootstrapResourceUrls", { enumerable: true, get: function () { return ui5_1.getBootstrapResourceUrls; } });
var vscode_helpers_1 = require("./vscode-helpers/vscode-helpers");
Object.defineProperty(exports, "getDefaultTargetFolder", { enumerable: true, get: function () { return vscode_helpers_1.getDefaultTargetFolder; } });
Object.defineProperty(exports, "isExtensionInstalled", { enumerable: true, get: function () { return vscode_helpers_1.isExtensionInstalled; } });
var app_gen_info_1 = require("./app-gen-info");
Object.defineProperty(exports, "generateAppGenInfo", { enumerable: true, get: function () { return app_gen_info_1.generateAppGenInfo; } });
var environment_1 = require("./environment");
Object.defineProperty(exports, "getHostEnvironment", { enumerable: true, get: function () { return environment_1.getHostEnvironment; } });
var app_helpers_1 = require("./app-helpers/app-helpers");
Object.defineProperty(exports, "getFlpId", { enumerable: true, get: function () { return app_helpers_1.getFlpId; } });
Object.defineProperty(exports, "getSemanticObject", { enumerable: true, get: function () { return app_helpers_1.getSemanticObject; } });
